### PR TITLE
jit: Introduce background compilation

### LINKED ISSUE #239: jit: Background compilation thread
In tiered compilation, the interpreter monitors the frequency of jumps to specific code addresses. Once a pre-set threshold is reached, it initiates JIT compilation. Concurrently, the emulator can continue operating with the interpreter, while JIT compilation is processed in a background thread, seamlessly switching to the JIT-compiled code once it is ready.

This emulator utilizes a JIT compiler for executing all RISC-V instructions. When execution begins, each emulator thread is equipped to compile new code. This implementation allows JIT compilation to occur in a background thread while the interpreter continues its execution until compilation is complete. If the emulator encounters a function that hasn't been compiled yet, it acquires a lock on the JIT code backend and attempts to compile the entire function into the JIT backend before resuming execution. Importantly, this lock only restricts other threads from adding new code to the JIT backend during compilation, without impeding their ability to use the JIT backend. Essentially, this means that one thread compiling new code has a minimal impact on other threads, resulting in a lock that imposes little overhead.

### PR DESCRIPTION
**Pull request by qwe661234**
Given the significant runtime compilation overhead associated with performing aggressive optimizations, we have implemented a background compilation mechanism to mitigate this issue. When the runtime profiler identifies a strong hotspot, it adds a T2C compilation request to the wait queue. A background thread, which continuously monitors this queue, triggers T2C to process the requests and notifies the main thread upon completion by updating a flag.

This mechanism defers the execution of T2C-generated machine code, leading to more frequent use of T1C-generated code. Despite this, the approach effectively minimizes runtime compilation delays by eliminating the need for the main thread to wait for T2C compilation to complete, thereby improving overall performance.

* The workflow of background compilation.
image

* The performance comparison with and without the background compialtion.
image

Close: #239

### CODE REVIEW

**Review by jserv — COMMENTED** [review]
# Benchmarks

<details>

| Benchmark suite | Current: 51ce53be52b626176e58bde93664c0d86989377a | Previous: 6b462c03dc8c64863ae146a94fe7abffe4da19bb | Ratio |
|-|-|-|-|
| `Dhrystone` | `5.22` Average DMIPS over 10 runs | `4.88` Average DMIPS over 10 runs | `0.93` |
| `Coremark` | `0.004` Average iterations/sec over 10 runs | `0.004` Average iterations/sec over 10 runs | `1` |

</details>

This comment was automatically generated by [workflow](https://github.com/sysprog21/rv32emu/actions?query=workflow%3ABenchmark) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).

File: src/riscv.c [hunk 1]
```diff
@@ -184,6 +186,31 @@ IO_HANDLER_IMPL(byte, write_b, W)
 #undef R
 #undef W
 
+#if RV32_HAS(JIT)
+static pthread_t t2c_thread;
+void *t2c_routine(void *arg)
```
Comment (jserv) [inline src/riscv.c]: Should be `static`.

File: src/riscv_private.h [hunk 2]
```diff
@@ -132,8 +140,11 @@ struct riscv_internal {
 #if !RV32_HAS(JIT)
     block_map_t block_map; /**< basic block map */
 #else
+    bool exit /**< Determine the main thread is exit or not */;
```
Comment (jserv) [inline src/riscv_private.h]: Rename to `quit` in order to distinguish from the state of being exited from the execution or program being terminated. In this case, using `quit` is much clearer.

File: src/riscv.c [hunk 3]
```diff
@@ -269,6 +296,10 @@ riscv_t *rv_create(riscv_user_t rv_attr)
     rv->jit_state = jit_state_init(CODE_CACHE_SIZE);
     rv->block_cache = cache_create(BLOCK_MAP_CAPACITY_BITS);
     assert(rv->block_cache);
+    rv->exit = false;
+    pthread_mutex_init(&rv->wait_queue_lock, NULL);
+    INIT_LIST_HEAD(&rv->wait_queue);
+    pthread_create(&t2c_thread, NULL, t2c_routine, rv);
```
Comment (jserv) [inline src/riscv.c]: Add some comments here. E.g., activate the background compilation thread.

File: src/riscv.c [hunk 4]
```diff
@@ -184,6 +186,31 @@ IO_HANDLER_IMPL(byte, write_b, W)
 #undef R
 #undef W
 
+#if RV32_HAS(JIT) && RV32_HAS(T2C)
+static pthread_t t2c_thread;
+static void *t2c_routine(void *arg)
```
Comment (jserv) [inline src/riscv.c]: Rename to `t2c_runloop` to emphasize its behavior in background thread.

**Comment by jserv** [comment]
I defer to @vacantron for confirmation.

File: src/riscv.c [hunk 5]
```diff
@@ -28,6 +28,8 @@
 #include "riscv_private.h"
 #include "utils.h"
 #if RV32_HAS(JIT)
+#include <pthread.h>
```
Comment (jserv) [inline src/riscv.c:32]: If T2C is not enabled, is pthread required?

File: src/riscv.c [hunk 6]
```diff
@@ -184,6 +186,31 @@ IO_HANDLER_IMPL(byte, write_b, W)
 #undef R
 #undef W
 
+#if RV32_HAS(JIT) && RV32_HAS(T2C)
```
Comment (jserv) [inline src/riscv.c]: We can centralize the dependency in `src/feature.h` to ensure T2C depends on JIT, eliminating the need to write `#if RV32_HAS(JIT) && RV32_HAS(T2C)`.

File: src/feature.h [hunk 7]
```diff
@@ -50,12 +50,11 @@
 /* Experimental just-in-time compiler */
 #ifndef RV32_FEATURE_JIT
 #define RV32_FEATURE_JIT 0
-#endif
-
 /* Experimental tier-2 just-in-time compiler */
 #ifndef RV32_FEATURE_T2C
 #define RV32_FEATURE_T2C 0
 #endif
+#endif
```
Comment (jserv) [inline src/feature.h:64]: Don't do that. Consider the following:
```c
#ifndef RV32_FEATURE_JIT
#define RV32_FEATURE_JIT 0
#endif

#ifndef RV32_FEATURE_T2C
#define RV32_FEATURE_T2C 0
#endif

/* T2C depends on JIT configuration */
#if !RV32_FEATURE_JIT
#undef RV32_FEATURE_T2C
#define RV32_FEATURE_T2C 0
#endif
```

File: src/riscv.c [hunk 8]
```diff
@@ -184,6 +187,31 @@ IO_HANDLER_IMPL(byte, write_b, W)
...
+            queue_entry_t *entry =
+                list_last_entry(&rv->wait_queue, queue_entry_t, list);
+            pthread_mutex_lock(&rv->wait_queue_lock);
+            list_del_init(&entry->list);
+            pthread_mutex_unlock(&rv->wait_queue_lock);
+            t2c_compile(entry->block,
+                        (uint64_t) ((memory_t *) PRIV(rv)->mem)->mem_base);
+            free(entry);
+        }
+        /* Instead of writing while(rv->quit), placing the code here prevents
```
Comment (jserv) [inline src/riscv.c]: Does `volatile` help for such purpose?

Comment (qwe661234) [reply src/riscv.c]: Yes, it works.

**Comment by jserv** [comment]
Finally, T2C is landed!

**Pull request merged**